feat(spooling): implement the Spooling protocol

fvaleye opened the pull request:
# Summary
This PR adds complete support for [Trino's spooling protocol](https://trino.io/docs/current/client/client-protocol.html#spooling-protocol), enabling handling of large query results by fetching data from external storage instead of embedding it directly in HTTP responses.

This PR has been tested locally with Docker:
- With Trino Cluster Version `478`
- With and without spooling enabled

### Key Features

- Automatic protocol detection: the client automatically detects and handles both `Direct` and `Spooled` protocols based on server responses
- Segment fetching: supports fetching both inline (base64-encoded) and remote (HTTP) segments with concurrent request handling (maximum concurrent requests could be set)
- Compression support: automatic decompression for `json`, `json+zstd`, and `json-lz4` formats based on Content-Encoding headers
- Segment acknowledgment: optional acknowledgment support for remote segments

### Implementation Details

- Added `spooling` optional feature gate with compression dependencies (`zstd`, `lz4`, `flate2`, `base64`)
- New module with spooling: `spooling::fetcher`, `spooling::decoder`, `spooling::encoding`, `spooling::segment`
-`QueryResultData` enum to support both Direct and Spooled variants
- Configurable via `ClientBuilder`: `spooling_encoding()` and `max_concurrent_segments()`: when the encoding is set, the header is added, and the server decides when it's best to use the spooling protocol

### Usage

```
let client = ClientBuilder::new("user", "localhost")
    .spooling_encoding("json+zstd")
    .max_concurrent_segments(10)  // Default based on CPU
    .build()?;
```

`let dataset = client.get_all::<MyStruct>(sql).await?;`
The client automatically uses the spooling protocol when the server supports it, with seamless fallback to the `Direct` protocol otherwise.

### Breaking Changes

None - this is a backward-compatible feature addition.

### Follow-ups
Use streaming/chunked processing for very large datasets

fvaleye committed:
feat(spooling): add the dependencies for spooling protocol as optional

Signed-off-by: Florian Valeye <florian.valeye@gmail.com>

fvaleye committed:
feat(spooling): add encoding and decoding for spooling protocols

Signed-off-by: Florian Valeye <florian.valeye@gmail.com>

fvaleye committed:
feat(spooling): implement Trino spooling protocol support

Add support for Trino's spooling protocol including segment fetching,
decompression, and automatic protocol detection. Supports both inline
and remote segments with configurable encoding and concurrency.

Signed-off-by: Florian Valeye <florian.valeye@gmail.com>

fvaleye committed:
feat(spooling): adapt the client to use Direct or Spooling depending on the server's responses

Added an example on how to use Spooling with the client

Signed-off-by: Florian Valeye <florian.valeye@gmail.com>

mfernd commented:
Thanks for the work @fvaleye! This looks great, we’ll review it shortly.

To fix the CI test failing, I think we should add these lines in the `Cargo.toml` for it to build with the `spooling` feature:
```toml
[[example]]
name = "spooling"
path = "examples/spooling.rs"
required-features = ["spooling"]

[[test]]
name = "spooling_protocol"
path = "tests/spooling_protocol.rs"
required-features = ["spooling"]
```

fvaleye commented:
> Thanks for the work @fvaleye! This looks great, we’ll review it shortly.
> 
> To fix the CI test failing, I think we should add these lines in the `Cargo.toml` for it to build with the `spooling` feature:
> 
> ```toml
> [[example]]
> name = "spooling"
> path = "examples/spooling.rs"
> required-features = ["spooling"]
> 
> [[test]]
> name = "spooling_protocol"
> path = "tests/spooling_protocol.rs"
> required-features = ["spooling"]
> ```

Thanks, I added the lines in the `Cargo.toml`

harksin commented on src/error.rs:
```diff
@@ -43,7 +43,7 @@ pub enum Error {
     HttpNotOk(StatusCode, String),
     #[error("query error, reason: {0}")]
-    QueryError(#[from] QueryError),
+    QueryError(Box<QueryError>),
```
why do we need to Box it ?

fvaleye replied:
It's to keep the Error enum size small.
`QueryError` is large (contains `Vec<String>` for stack traces).

harksin reviewed (approved):
lgtm, could be nice to introduce an integration test if possible before merging to test this e2e

fvaleye commented:
> lgtm, could be nice to introduce an integration test if possible before merging to test this e2e

Agreed! If I have time next week, I could integrate an integration test in a follow-up PR.

fvaleye committed:
chore(cargo): centralize Cargo configuration in workspace

Signed-off-by: Florian Valeye <florian.valeye@gmail.com>

fvaleye committed:
chore(test): add Docker setup for Trino integration tests

- Set up Trino server with a single coordinator and spooling protocol enabled.
- Include MinIO as intermediate storage for segment retrieval via spooling protocol.
- Recreate the environment on each startup using Docker Compose commands: stop, remove, and start.

Signed-off-by: Florian Valeye <florian.valeye@gmail.com>

fvaleye committed:
chore(test): add integration test to validate the implementation of Spooling protocol

Signed-off-by: Florian Valeye <florian.valeye@gmail.com>

fvaleye commented:
@harksin I added an integration test using a Docker-based Trino cluster with spooling enabled (and forced on small datasets). 
Also, I took the opportunity to centralize the Cargo configuration in the workspace.

fvaleye commented:
@harksin could you please approve the workflow to see if everything is green on the CI with the new integration test? Thanks!

harksin commented:
building :)

harksin commented:
@fvaleye released in version 0.9.2 🙂
https://crates.io/crates/trino-rust-client

Thanks a lot for your contribution :heart: !

And feel free to join our Discord if you want to chat: https://discord.com/invite/uk8XD9MGbh

(or if we take too long to review your next contribution ^^)

fvaleye commented:
> @fvaleye released in version 0.9.2 🙂 https://crates.io/crates/trino-rust-client
> 
> Thanks a lot for your contribution ❤️ !
> 
> And feel free to join our Discord if you want to chat: https://discord.com/invite/uk8XD9MGbh
> 
> (or if we take too long to review your next contribution ^^)

🚀 
With pleasure, thank you @harksin
